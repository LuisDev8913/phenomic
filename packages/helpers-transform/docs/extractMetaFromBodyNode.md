---
title: extractMetaFromBodyNode
---

@todo

* https://github.com/phenomic/phenomic/tree/master/packages/plugin-transform-markdown/src/index.js
* https://github.com/phenomic/phenomic/tree/master/packages/plugin-transform-asciidoc/src/index.js
