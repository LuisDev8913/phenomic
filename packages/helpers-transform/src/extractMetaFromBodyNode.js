// @flow

type Node =
  | string
  | {|
      t?: string,
      p?: Object,
      c?: Node | $ReadOnlyArray<Node>
    |};

const renderText = (node?: Node) => {
  if (!node) return "";
  if (typeof node === "string") return node;
  return Array.isArray(node.c)
    ? node.c.map((child: Node) => renderText(child)).join("")
    : renderText(node.c);
};

const getHeaders = (node?: Node) => {
  if (!node) return [];
  if (typeof node.t === "string") {
    const tag = node.t;
    const level = parseInt(tag[1], 10);
    if (tag[0] === "h" && !isNaN(level)) {
      return [[level, renderText(node)]];
    }
  }
  return (Array.isArray(node.c)
    ? // $FlowFixMe stfu
      node.c.reduce((acc, child: Node) => acc.concat(getHeaders(child)), [])
    : // $FlowFixMe stfu
      getHeaders(node.c)
  ).filter(h => h);
};

export default (node: Node) => {
  const headers = getHeaders(node);
  const firstH1 = headers.find(h => h[0] === 1);
  return {
    ...(firstH1 ? { title: firstH1[1] } : {}),
    ...(headers.length > 0 ? { headers } : {})
  };
};
