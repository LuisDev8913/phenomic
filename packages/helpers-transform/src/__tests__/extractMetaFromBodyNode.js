// @flow

import extractMetaFromBodyNode from "../extractMetaFromBodyNode.js";

it("should guess partial data from body AST", () => {
  expect(
    extractMetaFromBodyNode({
      c: [
        {
          c: [
            {
              c: [
                {
                  p: {
                    className: "icon icon-link"
                  },
                  t: "span"
                }
              ],
              p: {
                "aria-hidden": true,
                href: "#auto-title"
              },
              t: "a"
            },
            "Auto title"
          ],
          p: {
            id: "auto-title"
          },
          t: "h1"
        },
        "\n",
        {
          c: [
            {
              c: [
                {
                  p: {
                    className: "icon icon-link"
                  },
                  t: "span"
                }
              ],
              p: {
                "aria-hidden": true,
                href: "#sub-title"
              },
              t: "a"
            },
            "Sub title"
          ],
          p: {
            id: "sub-title"
          },
          t: "h2"
        },
        "\n",
        {
          c: ["Content"],
          t: "p"
        },
        "\n",
        {
          c: [
            {
              c: [
                {
                  p: {
                    className: "icon icon-link"
                  },
                  t: "span"
                }
              ],
              p: {
                "aria-hidden": true,
                href: "#another-level-2"
              },
              t: "a"
            },
            "Another Level 2"
          ],
          p: {
            id: "another-level-2"
          },
          t: "h2"
        },
        "\n",
        {
          c: [
            {
              c: [
                {
                  p: {
                    className: "icon icon-link"
                  },
                  t: "span"
                }
              ],
              p: {
                "aria-hidden": true,
                href: "#a-level-3"
              },
              t: "a"
            },
            "A level 3"
          ],
          p: {
            id: "a-level-3"
          },
          t: "h3"
        }
      ],
      t: "div"
    })
  ).toMatchSnapshot();
});
