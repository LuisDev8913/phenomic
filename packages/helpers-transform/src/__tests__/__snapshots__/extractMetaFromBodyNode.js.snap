// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should guess partial data from body AST 1`] = `
Object {
  "headings": Array [
    Object {
      "id": "auto-title",
      "level": 1,
      "text": "Auto title",
    },
    Object {
      "id": "sub-title",
      "level": 2,
      "text": "Sub title",
    },
    Object {
      "id": "another-level-2",
      "level": 2,
      "text": "Another Level 2",
    },
    Object {
      "id": "a-level-3",
      "level": 3,
      "text": "A level 3",
    },
  ],
  "title": "Auto title",
}
`;
